actionpack/lib/action_dispatch/request/session.rb

```diff
@@ -241,6 +242,11 @@ def each(&block)
         to_hash.each(&block)
       end
 
+      def initially_empty?
+        load_for_read!
+        @initially_empty ||= @delegate.empty?
+      end
```

Comment:
I think rather than `initially_empty?` I'd prefer something like `session.id_was`. It would be much simpler both here and in the CSRF store.

Reply:
(I haven't read the diff yet, but) that also seems relevant when processing a request that involves a change of session id -- in theory, we should always validate the incoming CSRF against the session id we started our request with, and always generate a CSRF token for the new/resulting session id.

(Consider e.g. a logout button that will clear the existing session [and change its id], but then also pushes "last page visited" value into the new session, for an after-re-login redirect.)

Reply:
It should be possible to capture the initial session ID, but I think that opens a whole other can of worms.

The cookie stores the CSRF token of the session at the time it is generated.  I believe that if the session does not exist, and has not previously been written to, this will be `nil` (the TestSession used in unit tests makes this a bit difficult to test), otherwise it will have a value.

If an `after_action` that resets the session runs after the token is generated, this is probably ok because it's equivalent to what we have today (i.e. the token is stored in the session, if the session gets reset, then the token disappears).

If an `after_action` that initializes the session runs after the token is generated, this will cause problems.  The cookie will indicate a session id of `nil`, but the session will actually have an ID.  This would cause validation to fail.  This isn't an issue with the existing code because the act of storing the token in the session initializes it.

Maybe this is a rare case, but it feels very similar to the `before_action` case we're trying to solve for now.  Am I over thinking this?

The best that I can think of is to set the cookie value to a lambda and update the cookie serializer to evaluate lambdas before serializing, which I don't think sounds like a great idea.